File: computer_sales.py
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
from scipy import stats
import pylab
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.api as sm
comp_sales=pd.read_csv("C:/Users/USER/Desktop/Computer_Data.csv")
comp_sales.columns
dum1=pd.get_dummies(comp_sales.cd)
dum1.columns="cd_no","cd_yes"
dum2=pd.get_dummies(comp_sales.multi)
dum2.columns="multi_no","multi_yes"
dum3=pd.get_dummies(comp_sales.premium)
dum3.columns="premium_no","premium_yes"
merged_comp=pd.concat([comp_sales,dum1,dum2,dum3],axis="columns")
merged_comp.columns
final=merged_comp.drop(['Unnamed: 0','cd','multi','premium'],axis="columns")

########univariet analysis#######
####price
plt.hist(final.price)
plt.boxplot(final.price)
measurements = np.random.normal(loc = 20, scale = 5, size=100)   
stats.probplot(final.price, dist="norm", plot=pylab)
pylab.show()
####speed
plt.hist(final.speed)
plt.boxplot(final.speed)
measurements = np.random.normal(loc = 20, scale = 5, size=100)   
stats.probplot(final.speed, dist="norm", plot=pylab)
pylab.show()
####ram
plt.hist(final.ram)
plt.boxplot(final.ram)
measurements = np.random.normal(loc = 20, scale = 5, size=100)   
stats.probplot(final.ram, dist="norm", plot=pylab)
pylab.show()
####hd
plt.hist(final.hd)
plt.boxplot(final.hd)
measurements = np.random.normal(loc = 20, scale = 5, size=100)   
stats.probplot(final.hd, dist="norm", plot=pylab)
pylab.show()

#########multi-variet analysis########
sns.pairplot(final)
np.corrcoef(final)
corr = final.dropna().corr()
corr
final.columns
#########
model=smf.ols('price~speed+hd+ram+screen+ads+trend+cd_no+multi_no+premium_no',data=final).fit()
model.summary()
sm.graphics.influence_plot(model)
plt.show()

#########
final1=final.drop(final.index[1440])
final2=final1.drop(final1.index[1700])
model1=smf.ols('price~speed+I(speed^2)+hd+I(hd^2)+ram+I(ram^2)+screen+I(screen^2)+ads+trend+cd_no+multi_no+premium_no',data=final2).fit()
model1.summary()

######train test error of model
model=smf.ols('price~speed+hd+ram+screen+ads+trend+cd_no+multi_no+premium_no',data=final).fit()
from sklearn.model_selection import train_test_split
final1_train,final1_test=train_test_split(final,test_size=0.2)

pred=model.predict(final1_train)
err=pred-final1_train.price
merr=np.mean(err*err)
rmse=np.sqrt(merr)
rmse

pred=model.predict(final1_test)
err=pred-final1_test.price
merr=np.mean(err*err)
rmse=np.sqrt(merr)
rmse
####train error=283 and test=271 and r2 value =77.4
######train test error of model1
model1=smf.ols('price~speed+I(speed^2)+hd+I(hd^2)+ram+I(ram^2)+screen+I(screen^2)+ads+trend+cd_no+multi_no+premium_no',data=final).fit()
from sklearn.model_selection import train_test_split
final1_train,final1_test=train_test_split(final,test_size=0.2)

pred=model1.predict(final1_train)
err=pred-final1_train.price
merr=np.mean(err*err)
rmse=np.sqrt(merr)
rmse

pred=model1.predict(final1_test)
err=pred-final1_test.price
merr=np.mean(err*err)
rmse=np.sqrt(merr)
rmse

######train-error=266.77 and test error=274.5###r2=78.6

File: computer_sales.R
library(readr)
library(plyr)
library(moments)
comp_sales<-read.csv(file.choose())
summary(comp_sales)
comp_sales$cd<- as.numeric(revalue(comp_sales$cd,c("yes"=1, "no"=0)))
comp_sales$premium<- as.numeric(revalue(comp_sales$premium,c("yes"=1, "no"=0)))
comp_sales$multi<- as.numeric(revalue(comp_sales$multi,c("yes"=1, "no"=0)))
str(comp_sales)
######univariet analysis
######price
hist(comp_sales$price,col="pink")
boxplot(comp_sales$price,col="pink",horizontal =T )
qqnorm(log(comp_sales$price),col="pink",pch=20)
qqline(log(comp_sales$price),col="pink")
skewness(comp_sales$price)
kurtosis(comp_sales$price)
########speed
hist(comp_sales$speed,col="pink")
boxplot(comp_sales$speed,col="pink",horizontal =T )
qqnorm(comp_sales$speed,col="pink",pch=20)
qqline(comp_sales$speed,col="pink")
skewness(comp_sales$speed)
kurtosis(comp_sales$speed)
########ram
hist(comp_sales$ram,col="pink")
boxplot(comp_sales$ram,col="pink",horizontal =T )
qqnorm(comp_sales$ram,col="pink",pch=20)
qqnorm(log(comp_sales$ram),col="pink",pch=20)
qqline(log(comp_sales$ram),col="pink")
skewness(comp_sales$ram)
kurtosis(comp_sales$ram)
pairs(comp_sales)
#########hd
hist(comp_sales$hd,col="pink")
boxplot(comp_sales$hd,col="pink",horizontal =T )
qqnorm(log(comp_sales$hd),col="pink",pch=20)
qqline(log(comp_sales$hd),col="pink")
skewness(comp_sales$hd)
kurtosis(comp_sales$hd)
###multi variet analysis
pairs(comp_sales,col="pink",pch=20)

library(corpcor)

cor2pcor(cor(comp_sales))
cor(comp_sales)
comp_sales1<-comp_sales[,-1]
model<-lm(price~.,data=comp_sales1)
summary(model)
plot(model)
library(car)
infIndexPlot(model, vars=c("Cook", "Studentized", "Bonf", "hat"),
             id=TRUE, grid=TRUE, main="Diagnostic Plots")
vif(model)
avPlots(model,id.n=2,id.cex=0.8,col="red")
##########1441,1701#####
model1<-lm(log(price)~.,data=comp_sales[-c(1441,1701),])

summary(model1)
plot(model1)
##########transformations##############
colnames(comp_sales)
model1<-lm(log(price)~speed+I(speed*speed)+hd+I(hd*hd)+ram+I(ram*ram)+screen+cd+multi+
             premium+ads+trend,data=comp_sales[-c(1441,1701),])
infIndexPlot(model1, vars=c("Cook", "Studentized", "Bonf", "hat"),
             id=TRUE, grid=TRUE, main="Diagnostic Plots")
summary(model1)
plot(model1,col="pink",pch=20)
vif(model1)
############train test###############
comp_sales1<-comp_sales[-c(1441,1701),-1]
n<-nrow(comp_sales)
n1<-n*0.8
n2<-n-n1
train<-sample(1:n,n1)
comp_train<-comp_sales[train,]
comp_test<-comp_sales[-train,]

model1<-lm(log(price)~speed+I(speed*speed)+hd+I(hd*hd)+ram+screen+I(screen*screen)+cd+multi+
             premium+ads+trend,data=comp_train)
pred<-predict(model1,newdata =comp_test,interval = "confidence")
pred_exp<-exp(pred)
pred_exp<-as.data.frame(pred_exp)
err<-pred_exp$fit-comp_test$price

rmse<-sqrt(mean(err*err))

pred<-predict(model1,newdata =comp_train,interval = "confidence")
pred_exp<-exp(pred)
pred_exp<-as.data.frame(pred_exp)
err<-pred_exp$fit-comp_train$price
rmse1<-sqrt(mean(err*err))
rmse1
############model1 train test################


###########model###########
comp_sales1<-comp_sales[-c(1441,1701),-1]
n<-nrow(comp_sales1)
n1<-n*0.8
n2<-n-n1
train<-sample(1:n,n1)
comp_train<-comp_sales1[train,]
comp_test<-comp_sales1[-train,]
model<-lm(log(price)~.,data=comp_sales1)
pred<-predict(model,newdata =comp_test,interval = "confidence")
pred_exp<-exp(pred)
pred_exp<-as.data.frame(pred_exp)
err<-pred_exp$fit-comp_test$price
rmse<-sqrt(mean(err*err))
rmse

pred<-predict(model,newdata =comp_train,interval = "confidence")
pred_exp<-exp(pred)
pred_exp<-as.data.frame(pred_exp)
err<-pred_exp$fit-comp_train$price

rmse1<-sqrt(mean(err*err))
rmse1
